### PR TITLE
[SPARK-25482][SQL] Avoid pushdown of subqueries to data source filters

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

An expressions with a subquery can be pushed down as a data source filter. Despite the filter is not actively used, this causes anyway a re-execution of the subquery becuase the `ReuseSubquery` optimization rule is ineffective in this case.

The PR avoids this problem by forbidding the push down of filters containing a subquery.
## How was this patch tested?

added UT

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala [hunk 1]
```diff
@@ -166,7 +168,7 @@ case class ReuseSubquery(conf: SQLConf) extends Rule[SparkPlan] {
         val sameSchema = subqueries.getOrElseUpdate(sub.plan.schema, ArrayBuffer[SubqueryExec]())
         val sameResult = sameSchema.find(_.sameResult(sub.plan))
         if (sameResult.isDefined) {
-          sub.withNewPlan(sameResult.get)
+          sub.withNewPlan(sameResult.get).withNewExprId()
```
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: I think it is cleaner this way. I don't expect this to happen very often (how many subquery can you have in a plan?) so I don't think it is an issue. But if there are cleaner options/solutions, I am open to suggestions, thanks.

**Comment by mgaido91** [comment]
cc @cloud-fan @dongjoon-hyun @gatorsmile

**Comment by cloud-fan** [comment]
> This can happen for instance when a filter containing a scalar subquery is pushed to a DataSource

hmm, how can this happen? I don't think a data source can handle a filter of subquery...

**Comment by mgaido91** [comment]
> hmm, how can this happen?

you can check the UT which reproduces the issue. The scalar subquery is pushed down as part of the filter `GreaterThen`

**Comment by gengliangwang** [comment]
@mgaido91 Sorry but can you have a more detailed explanation in the PR description? 
With the code changes, the predicate with subquery result can be push down into data source. Is this the main point of the PR? And why is that creating a new expr ID can fix it?

**Comment by mgaido91** [comment]
@gengliangwang no, let me cite and explain the PR description. I am not sure how to improve it, but if you have suggestions I am happy to. The main point of the PR is to address an issue which arise when:

> When a `ExecSubqueryExpression` is copied

Now the point is, can this condition happen? The answer is yes, and one situation in which this happens (as reported in the JIRA) is

> when a filter containing a scalar subquery is pushed to a DataSource.

So in the plan we have two `ExecSubqueryExpression` each with a copy of the same `SubqueryExec`. The problem which arises in this condition is that:

> `ReuseSubquery` becomes useless, as replacing the `SubqueryExec` is ignored since the new plan is equal to the previous one.

So this result in the subquery being executed twice (as the two `SubqueryExec` are distinct, despite they are the same).

**Comment by gengliangwang** [comment]
@mgaido91 I see, thanks for the explanation!

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 2]
```diff
@@ -1268,4 +1269,16 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       assert(getNumSortsInQuery(query5) == 1)
     }
   }
+
+  test("SPARK-25482: Reuse same Subquery in order to execute it only once") {
+    withTempView("t1", "t2") {
+      sql("create temporary view t1(a int) using parquet")
+      sql("create temporary view t2(b int) using parquet")
+      val plan = sql("select * from t2 where b > (select max(a) from t1)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: sorry it has been a long time and I don't quite remember the context.

What was the problem we are trying to fix? This test looks nothing related to subquery reuse.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Sure, please can you check the PR description? I think the context is quite well explained there.

Anyway, as a quick summary: in this case `b > (select max(a) from t1)` is pushed down as a datasource filter. So we have 2 instances of `b > (select max(a) from t1)` and the result is not reused. It is not reused because the copied plan satisfies `==`, so even if `ReuseSubquery` replaces it, then the change is ignored.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Do we only have a problem when we have subquery in data source filter? If that's the case, I would suggest not pushdown subquery filter into data source.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: we have this problem if we `copy` the same subquery. I can't think of any other case than filter push-down, but I may be wrong.

Forbidding to push down these filter may cause a perf regression, I am not sure it is the right solution.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: is there any data source can support subquery filter? for data source v1/v2, the public `Filter` API does not support subquery. For file source, they don't support subquery filter either.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: I see your point. I'd argue that scalar subqueries like this may be used for Filter pushdown: since the subquery is evaluated before the plan execution, their result _may_ be used in the filter push down. With the current codebase you're 100% right and I can follow your comment/suggestion. But wouldn't it be worth to let them be used (and do that in another PR)?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: I think you are right about it, but it also means the data source scan must wait until the subquery is finished. We need to make tradeoffs carefully.

I'd suggest we open a new ticket about scalar subquery filter pushdown to data source, and forbid it here.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: > The subquery should be executed anyway sooner or later, right?

Yes, but we could execute scan and subquery at the same time (2 spark jobs running together), instead of executing them serialized.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: > we could execute scan and subquery at the same time (

is this really possible? My understanding is that subqueries are executed before the plan they belong to (in `SparkPlan.executeQuery`). So my understanding is that when a subquery is running, the rest of the query is not.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: ah sorry I misread the code. Unless the subquery is rewritten into join, we must wait for all subqueries to be finished before executing the plan.

We can rewrite scalar subquery in data source filters into literal, to make it work with the filter pushdown API.

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: yes, exactly. That's what I meant. Shall I revert the changes to the previous PR? And in the scope of the new JIRA do the rewriting to literals? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 3]
```diff
@@ -155,15 +155,14 @@ object FileSourceStrategy extends Strategy with Logging {
           case a: AttributeReference =>
             a.withName(l.output.find(_.semanticEquals(a)).get.name)
         }
-      }
+      }.filterNot(SubqueryExpression.hasSubquery)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: shall we do the filter before the `map`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala [hunk 4]
```diff
@@ -47,7 +47,8 @@ private[sql] object PruneFileSourcePartitions extends Rule[LogicalPlan] {
           case a: AttributeReference =>
             a.withName(logicalRelation.output.find(_.semanticEquals(a)).get.name)
         }
-      }
+      }.filterNot(SubqueryExpression.hasSubquery)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 5]
```diff
@@ -1268,4 +1269,16 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       assert(getNumSortsInQuery(query5) == 1)
     }
   }
+
+  test("SPARK-25482: Reuse same Subquery in order to execute it only once") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: let's update the test

**Comment by cloud-fan** [comment]
I'd like to merge this simple PR first, to address the performance problem (unnecessary subquery execution).

Let's create a new ticket for subquery filter pushing to data source, and have more people to attend the discussion.

**Comment by cloud-fan** [comment]
BTW can you include a simple benchmark to show this problem? e.g. just run a query in spark-shell, and post the result before and after this PR.

**Comment by mgaido91** [comment]
@cloud-fan this is the benchmark:
```
(1 to 1000000).toSeq.toDF("a").write.save("/tmp/t1")
spark.read.load("/tmp/t1").createTempView("t1")
(1 to 2000).toSeq.toDF("b").write.save("/tmp/t2")
spark.read.load("/tmp/t2").createTempView("t2")
val plan = sql("select * from t2 where b > (select avg(a + 1) from t1)")
val t0 = System.nanoTime()
plan.show
val t1 = System.nanoTime()
println("Elapsed time: " + (t1 - t0) + "ns")
```

the result is:

```
Before PR: Elapsed time: 862499689ns
After  PR: Elapsed time: 914728641ns
```
The difference is very small because all the subqueries run in parallel. The execution time would be much more affected if there are several subqueries (our thread pool has 16 threads so a query like that but with 9 filters with subqueries would see a big performance gain after this PR).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala [hunk 6]
```diff
@@ -22,7 +22,7 @@ import scala.collection.mutable.ArrayBuffer
 
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.{expressions, InternalRow}
-import org.apache.spark.sql.catalyst.expressions.{Expression, ExprId, InSet, Literal, PlanExpression}
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExprId, InSet, Literal, NamedExpression, PlanExpression}
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 7]
```diff
@@ -1268,4 +1269,16 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
       assert(getNumSortsInQuery(query5) == 1)
     }
   }
+
+  test("SPARK-25482: Forbid pushdown to dattasources of filters containing subqueries") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: `dattasources` typo

**Comment by cloud-fan** [comment]
thanks, merging to master!
